app/models/mediaflux/http/request.rb

```diff
@@ -60,27 +60,40 @@ def self.default_xml_namespace_uri
         "http://tigerdata.princeton.edu"
       end
 
+      def self.find_or_create_http_client
+        @http_client ||= begin
+                           @http_client = Net::HTTP::Persistent.new
+                           # https is not working correctly on td-meta1 we should not need this, but we do...
+                           @http_client.verify_mode = OpenSSL::SSL::VERIFY_NONE
+                           @http_client
+                         end
+      end
+
+      def self.create_finalizer
+        proc {
+          @http_client.shutdown unless @http_client
+        }
+      end
+
       attr_reader :session_token
 
       # Constructor
       # @param file [File] any upload file required for the POST request
       # @param session_token [String] the API token for the authenticated session
       # @param http_client [Net::HTTP::Persistent] HTTP client for transmitting requests to the Mediaflux server API
       def initialize(file: nil, session_token: nil, http_client: nil)
-        @http_client = http_client || Net::HTTP::Persistent.new
-        # https is not working correctly on td-meta1 we should not need this, but we do...
-        @http_client.verify_mode = OpenSSL::SSL::VERIFY_NONE
-
+        @http_client = http_client || self.class.find_or_create_http_client
         @file = file
         @session_token = session_token
+
+        finalizer = self.class.create_finalizer
+        ObjectSpace.define_finalizer(self, finalizer)
```

Comment:
This implements the following pattern:

```ruby
class Foo
  def initialize(data_needed_for_finalization)
    ObjectSpace.define_finalizer(self, self.class.create_finalizer(data_needed_for_finalization))
  end

  def self.create_finalizer(data_needed_for_finalization)
    proc {
      puts "finalizing #{data_needed_for_finalization}"
    }
  end
end
```

https://ruby-doc.org/3.3.0/ObjectSpace.html#method-c-define_finalizer

Reply:
I now see two errors here, I am very sorry.